package_scanner.nim

```diff
@@ -1,135 +1,367 @@
+  packagesJsonStr = readFile(packagesFilePath)
+  hostsSkip = [
+    "https://bitbucket",
+    "https://gitlab.3dicc",
+    "https://mahlon@bitbucket",
+    "https://notabug",
+  ] ## Hostnames to skip checks, for reliability, wont like direct HTTP GET?.
+  licenses = [
+    "allegro 4 giftware",
+    "apache",
+    "apache2",
+    "apache2.0",
+    "apache 2.0",
+    "apache-2.0",
+    "apache version 2.0",
+    "apache license 2.0",
+    "bsd",
+    "bsd2",
+    "bsd-2",
+    "bsd3",
+    "bsd-3",
+    "bsd 2-clause",
+    "bsd-2-clause",
+    "bsd 3-clause",
+    "bsd-3-clause",
+    "cc0",
+    "gpl",
+    "gpl2",
+    "gpl-2.0",
+    "gpl3",
+    "gpl-3.0",
+    "gplv2",
+    "gplv3",
+    "lgpl",
+    "lgplv2",
+    "lgplv2.1",
+    "lgplv3",
+    "mit",
+    "ms-pl",
+    "mpl",
+    "ppl",
+    "wtfpl",
     "libpng",
     "zlib",
-    "ISC",
-    "Unlicense"
-  ]
-  VCS_TYPES = @["git", "hg"]
+    "isc",
+    "unlicense",
+    # Not concrete Licenses, but some rare cases observed on the JSON.
+    "0bsd",
+    "agplv3",
+    "agpl-3.0",
+    "apache license 2.0 or mit",
+    "fontconfig",
+    "gnu lesser general public license v2.1",
+    "public domain",
+    "lgplv3 or gplv2",
+    "openssl and ssleay",
+    "openldap",
+    "apache 2.0 or gplv2",
+    "mit or apache 2.0",
+    "lgpl with static linking exception",
+  ]  ## All valid known licences for Nimble packages, on lowercase.
+when checkProfanity:
+  const profanitiesStr = readFile(currentSourcePath().parentDir / "profanities.txt").normalize.splitLines
+
+let
+  packagesJson = parseJson(packagesJsonStr).getElems         ## ``string`` to ``JsonNode``
+  pckgsList = filterIt(packagesJson, not it.hasKey("alias")) ## Packages, without Aliases
+  aliasList = filterIt(packagesJson, it.hasKey("alias"))     ## Aliases, without Packages
+  client = newHttpClient(timeout = httpTimeout) ## HTTP Client with Timeout
+  report = newJUnitOutputFormatter(openFileStream("report.xml", fmWrite)) ## JUnit Report XML
+addOutputFormatter(defaultConsoleFormatter())
+addOutputFormatter(report)
+when checkProfanity:
+  let profanities = filterIt(profanitiesStr, it.len > 0 and not(it.startsWith"#"))
+
+proc handler() {.noconv.} =
+  quit("CTRL+C Pressed, package_scanner is shutting down, Bye.")
+setControlCHook(handler)
+
+using url, name: string
 
-proc canFetchNimbleRepository(name: string, urlJson: JsonNode): bool =
-  # The fetch is a lie!
-  # TODO: Make this check the actual repo url and check if there is a
-  #       nimble file in it
-  result = true
-  var url: string
+func stripEndingSlash(url): string {.inline.} =
+  ## Strip the ending '/' if any else return the same string.
+  result = if url[^1] == '/': url[0 .. url.len - 2] else: url
 
-  if not urlJson.isNil:
-    url = urlJson.str
+func preprocessUrl(url, name): string =
+  ## Take **Normalized** URL & Name return Download link. GitHub & GitLab supported.
+  if likely(url.startswith("https://github.com/")):
```

Comment:
The tool will spend most time doing networking and little else. Optimizing for branch prediction sounds quite unnecessary :)